resourceIndexer/core-url-remapping.js

```diff
@@ -0,0 +1,49 @@
+import { readFileSync } from "fs";
+import { convertManifestJsonToAvailableDatasetList } from "../src/endpoints/charon/setAvailableDatasets.js";
+
+let resolvedPath;
+
+/**
+ * Remaps an incomplete URL core dataset path to the path it will be redirected
+ * to by the server via the "nextstrain manifest". If the provided URL would not
+ * be redirected we return the provided path unchanged.
+ * @param {string} path nextstrain.org URL path
+ * @returns {string}
+ */
+export const remapCoreUrl = (path) => {
+  if (!resolvedPath) {
+    /* map only computed the first time this function is run */
+    const serverManifest = JSON.parse(readFileSync("./data/manifest_core.json"));
+    resolvedPath = resolveAll(
+      convertManifestJsonToAvailableDatasetList(serverManifest).defaults
+    )
+  }
+  return resolvedPath[path] || path;
+}
+
+
+/**
+ * The server contains a map of partial paths and their next (default) path,
+ * e.g. flu → seasonal
+ *      flu/seasonal → h3n2
+ *      etc
+ * and uses an iterative approach to resolve the final path. Here we want to
+ * resolve the entire path for every partial path, for instance
+ *      flu → flu/seasonal/h3n2/h2/2y
+ *      flu/seasonal → flu/seasonal/h3n2/h2/2y
+ * @param {Object} defaults 
+ * @returns {Object}
+ */
+function resolveAll(defaults) {
+  const urlMap = {};
+  const defaultKeys = new Set(Object.keys(defaults));
+  Object.entries(defaults).forEach(([partialPath, nextPathPart]) => {
+    if (partialPath==='') return; // manifest has a base default of zika!
+    let resolvedPath = `${partialPath}/${nextPathPart}`;
+    while (defaultKeys.has(resolvedPath)) {
+      resolvedPath += `/${defaults[resolvedPath]}`;
+    }
```

Comment:
I see this sort of dual variable separation of the keys (in `defaultKeys`) from the actual values (in `defaults`)—where you check variable A but then access variable B—to be an anti-pattern. If what you're reaching for is the `.has()` method, you can use a Map instead and keep the data in a single variable, e.g.:

```js
const defaultPaths = new Map(Object.entries(defaults));
// …
  while (defaultPaths.has(resolvedPath)) {
    resolvedPath += `/${defaultPaths.get(resolvedPath)}`;
  }
```

You can also directly iterate over a Map, so combined with my `for (…) {` suggestion above, you'd do:

```js
for (const [partialPath, nextPathPart] of defaultPaths) {
```

Reply:
Ok. Changed to checking membership of `defaults` without maintaining a parallel `defaultKeys`.